Adding OS tracking support

bkus opened the pull request:
WARNING: Not tested since I don't know how to test portal.  It'll also need a migration applied before deploying.

kd7lxl commented on portal/models.py:
```diff
@@ -72,6 +80,7 @@ class Host(models.Model):
     name = models.CharField(max_length=242, unique=True,
         validators=[domain_validator])
     type = models.CharField(max_length=24, choices=HOST_TYPES)
+    os = models.CharField(max_length=32, choices=OS_TYPES)
```
Please use `null=True` so that we don't need to identify all of the existing hosts in the migration. This will allow OS to be NULL until you try to edit the host in the admin interface.

bkus replied:
Done.

kd7lxl commented on portal/models.py:
```diff
@@ -37,6 +37,14 @@
+    ('esxi', 'ESXi'),
+    ('windows', 'Windows'),
+    ('ilo', 'iLO'),
+    ('other', 'Other'),
```
Previously we had `None` type for hosts where OS is not applicable, like anycast services. How does that translate here?

bkus replied:
I've added a (None, 'None') OS choice, which should translate to NULL in SQL.

kd7lxl commented on portal/views.py:
```diff
@@ -34,20 +34,6 @@ def all_hosts(request):
 def ansible_hosts(request):
     """Generates Ansible dynamic inventory format"""
 
-    # hint OS by assuming all routers are Mikrotik
-    os = {
-        'sector': 'mikrotik',
-        'ptp': 'mikrotik',
-        'edgerouter': 'mikrotik',
-        'cellrouter': 'mikrotik',
-        'client': 'mikrotik',
-        'server': 'linux',
-        'anycast': None,
-        'pdu': None,
-        'kvm': None,
-        'other': 'linux',
-    }
```
Please make this backwards compatible so that we still have a way to target all the Mikrotik devices. Maybe preserve this default OS identification for when OS is NULL.

bkus replied:
This will be handled in a post-migration SQL action.  'routeros' will be set for all the 'sector', 'ptp', 'edgerouter', 'cellrouter', 'client' hosts.

kd7lxl replied:
This might be what you meant by "post-migration", but a manually-written line near the end of the generated migration could handle this.

bkus replied:
Intended SQL:

UPDATE portal_host SET os='routeros' WHERE type IN ('sector', 'ptp', 'edgerouter', 'cellrouter', 'client');

kd7lxl reviewed (commented):
Check the Django docs to learn how to make migrations. The migration should be part of the same PR.

bkus commented:
> Check the Django docs to learn how to make migrations. The migration should be part of the same PR.

Tried to virtualenv on Seattle-SRV1:

```
(env)eo@Seattle-SRV1:~/clean/hamwan-portal$ pip install -r requirements.txt 
Downloading/unpacking Django>=1.7,<1.8 (from -r requirements.txt (line 1))
  Cannot fetch index base URL http://pypi.python.org/simple/
  Could not find any downloads that satisfy the requirement Django>=1.7,<1.8 (from -r requirements.txt (line 1))
No distributions at all found for Django>=1.7,<1.8 (from -r requirements.txt (line 1))
Storing complete log in /home/eo/.pip/pip.log
```

The log complains about "must use HTTPS", so I did:

```
(env)eo@Seattle-SRV1:~/clean/hamwan-portal$ pip install --index-url=https://pypi.python.org/simple/ -r requirements.tx
t 
Downloading/unpacking Django>=1.7,<1.8 (from -r requirements.txt (line 1))
  Cannot fetch index base URL https://pypi.python.org/simple/
  Could not find any downloads that satisfy the requirement Django>=1.7,<1.8 (from -r requirements.txt (line 1))
No distributions at all found for Django>=1.7,<1.8 (from -r requirements.txt (line 1))
Storing complete log in /home/eo/.pip/pip.log
```

The log now complaints that HTTPS must use SNI:

```
Downloading/unpacking Django>=1.7,<1.8 (from -r requirements.txt (line 1))

  Getting page https://pypi.python.org/simple/Django
  Could not fetch URL https://pypi.python.org/simple/Django: HTTP Error 403: SNI is required
  Will skip URL https://pypi.python.org/simple/Django when looking for download links for Django>=1.7,<1.8 (from -r requirements.txt (line 1))
  Getting page https://pypi.python.org/simple/
  Could not fetch URL https://pypi.python.org/simple/: HTTP Error 403: SNI is required
  Will skip URL https://pypi.python.org/simple/ when looking for download links for Django>=1.7,<1.8 (from -r requirements.txt (line 1))
  Cannot fetch index base URL https://pypi.python.org/simple/

  URLs to search for versions for Django>=1.7,<1.8 (from -r requirements.txt (line 1)):
  * https://pypi.python.org/simple/Django/
  Getting page https://pypi.python.org/simple/Django/
  Could not fetch URL https://pypi.python.org/simple/Django/: HTTP Error 403: SNI is required
  Will skip URL https://pypi.python.org/simple/Django/ when looking for download links for Django>=1.7,<1.8 (from -r requirements.txt (line 1))
  Could not find any downloads that satisfy the requirement Django>=1.7,<1.8 (from -r requirements.txt (line 1))

No distributions at all found for Django>=1.7,<1.8 (from -r requirements.txt (line 1))
```

Then I tried to do this on my modern box, with "python3-virtualenv" (the only one available):

```
(env) [eo@ansible hamwan-portal]$ pip install -r requirements.txt 
Collecting Django<1.8,>=1.7
  Downloading Django-1.7.11-py2.py3-none-any.whl (7.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 13.3 MB/s eta 0:00:00
Collecting argparse==1.2.1
  Downloading argparse-1.2.1.tar.gz (69 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.3/69.3 kB 2.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting djangorestframework==3.9.1
  Downloading djangorestframework-3.9.1-py2.py3-none-any.whl (950 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 950.8/950.8 kB 28.8 MB/s eta 0:00:00
Collecting ipaddr==2.1.11
  Downloading ipaddr-2.1.11.tar.gz (29 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [9 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-jcalomy1/ipaddr_ce316c4ed6bc40ccbdb246f7b175cc4b/setup.py", line 19, in <module>
          import ipaddr
        File "/tmp/pip-install-jcalomy1/ipaddr_ce316c4ed6bc40ccbdb246f7b175cc4b/ipaddr.py", line 1450
          ip_int = 0L
                   ^
      SyntaxError: invalid decimal literal
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

[notice] A new release of pip available: 22.2.2 -> 22.3.1
[notice] To update, run: pip install --upgrade pip
(env) [eo@ansible hamwan-portal]$
```

At least it's new enough to talk to the internet!  But...any ideas how to resolve that problem?

I seem unable to create a virtualenv, and therefore migration.

kd7lxl commented:
You're way ahead of yourself here. You should be generating the migration and testing it locally before trying on Seattle-SRV1. Check out https://docs.djangoproject.com/en/1.8/topics/migrations/. Also https://docs.djangoproject.com/en/1.8/intro/tutorial01/#the-development-server.

bkus commented:
> You're way ahead of yourself here. You should be generating the migration and testing it locally before trying on Seattle-SRV1. Check out https://docs.djangoproject.com/en/1.8/topics/migrations/. Also https://docs.djangoproject.com/en/1.8/intro/tutorial01/#the-development-server.

Doesn't that all need to happen from within the virtualenv?  That's what I can't get going in either location.

kd7lxl commented:
Hmm, it's possible the referenced dependencies no longer exist in pypi. I've never had to solve this problem. Should have containerized it I guess.

bkus commented:
The deps seem to exist, but one of them has a parsing problem.  See the "invalid decimal literal" problem?  Maybe we can bump the ipaddr version in requirements.txt?

bkus commented:
Yup, removing the version req for ipaddr got it to install.  Now the problem is with the last dep, wsgiref:

```
Collecting wsgiref
  Using cached wsgiref-0.1.2.zip (37 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [9 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-u_e9f9_7/wsgiref_9f7f2548b6c7456198888a1957f1cb1c/setup.py", line 5, in <module>
          import ez_setup
        File "/tmp/pip-install-u_e9f9_7/wsgiref_9f7f2548b6c7456198888a1957f1cb1c/ez_setup/__init__.py", line 170
          print "Setuptools version",version,"or greater has been installed."
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
      [end of output]
```

There isn't a newer version for that, and it's quite old.  I see django_wsgi is newer, perhaps worth porting the portal over to that?

bkus commented:
I removed the uwsgiref requirement for now since it should only be required to run the server, not to makemigrations.  pip install -r requirements.txt worked.  However, django is still not available?

```
(env) [eo@ansible hamwan-portal]$ vim requirements.txt 
(env) [eo@ansible hamwan-portal]$ pip install -r requirements.txt 
Collecting Django<1.8,>=1.7
  Using cached Django-1.7.11-py2.py3-none-any.whl (7.4 MB)
Collecting argparse==1.2.1
  Using cached argparse-1.2.1.tar.gz (69 kB)
  Preparing metadata (setup.py) ... done
Collecting djangorestframework==3.9.1
  Using cached djangorestframework-3.9.1-py2.py3-none-any.whl (950 kB)
Collecting ipaddr
  Using cached ipaddr-2.2.0.tar.gz (26 kB)
  Preparing metadata (setup.py) ... done
Collecting psycopg2-binary>=2.6
  Using cached psycopg2_binary-2.9.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.0 MB)
Collecting pyasn1==0.1.7
  Using cached pyasn1-0.1.7.tar.gz (68 kB)
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: argparse, pyasn1, ipaddr
  Building wheel for argparse (setup.py) ... done
  Created wheel for argparse: filename=argparse-1.2.1-py3-none-any.whl size=21381 sha256=fdbb0dc5f590470bab82cddbdd0ca483d0865e93cdad0899bdba29831bf8442f
  Stored in directory: /home/eo/.cache/pip/wheels/ea/f4/a2/fa174a3c5c9e4637dfa962fd2bb39c50085eaad4c8e0f3ecbc
  Building wheel for pyasn1 (setup.py) ... done
  Created wheel for pyasn1: filename=pyasn1-0.1.7-py3-none-any.whl size=32821 sha256=d6f035abec898cdd36842d64a5563d0f50a566c2987b5815bd41c3cdae3d4629
  Stored in directory: /home/eo/.cache/pip/wheels/36/83/d2/66dfb148de3891e1b7da1011a5822c5f42938a5bbf1ae4eb3e
  Building wheel for ipaddr (setup.py) ... done
  Created wheel for ipaddr: filename=ipaddr-2.2.0-py3-none-any.whl size=18281 sha256=3d8beaa54378adc4515626e0dbc3cee69e3a6b6fb1ef204667cc8a9ac7073761
  Stored in directory: /home/eo/.cache/pip/wheels/1b/f4/54/ce8b4b51b444da2d12c64f5a99b6e4b526f341f803d4c9f5ac
Successfully built argparse pyasn1 ipaddr
Installing collected packages: pyasn1, ipaddr, Django, argparse, psycopg2-binary, djangorestframework
Successfully installed Django-1.7.11 argparse-1.2.1 djangorestframework-3.9.1 ipaddr-2.2.0 psycopg2-binary-2.9.5 pyasn1-0.1.7

[notice] A new release of pip available: 22.2.2 -> 22.3.1
[notice] To update, run: pip install --upgrade pip
(env) [eo@ansible hamwan-portal]$ python2.7 manage.py makemigrations
Traceback (most recent call last):
  File "manage.py", line 8, in <module>
    from django.core.management import execute_from_command_line
ImportError: No module named django.core.management
(env) [eo@ansible hamwan-portal]$ 
```
I added a django-core requirement, and it installed fine too, but same error persists.  :(

bkus commented:
Problem was with virtualenv being python3 now.  Solved with virtualenv -p /usr/bin/python2.7 env.

kd7lxl commented on portal/migrations/0002_host_os.py:
```diff
@@ -0,0 +1,20 @@
+            field=models.CharField(max_length=32, null=True, choices=[(b'routeros', b'RouterOS'), (b'airos', b'AirOS'), (b'linux', b'Linux'), (b'esxi', b'ESXi'), (b'windows', b'Windows'), (b'ilo', b'iLO'), (b'other', b'Other'), (None, b'None')]),
+            preserve_default=True,
+        ),
+    ]
```
Following the docs, you should be able to find an example of how to set initial values for the routers somewhere around this part of the migration.

bkus replied:
Here's the command:

UPDATE portal_host SET os='routeros' WHERE type IN ('sector', 'ptp', 'edgerouter', 'cellrouter', 'client');

Please provide Django translation and I'll update PR.

kd7lxl replied:
Check out https://docs.djangoproject.com/en/1.8/topics/migrations/#data-migrations. You will have to write some python.

bkus replied:
Nah, I'm not gonna waste time re-expressing perfectly good SQL in Python.  My time is valuable, and this isn't even the main quest.  Take it or leave it.

kd7lxl replied:
I'm sure there is a way to pass raw SQL through the ORM in the migration, but since that's not the normal way of doing it I can't offer any expertise.

I noticed the SQL you provided is not idempotent if you happened to run the migration again after manually managing some records. I'd recommend introducing a WHERE NOT NULL clause, regardless of which language you write the migration in.

bkus replied:
It's ephemeral SQL, I'm not worried about idempotency.  Maybe at some point later you can share the expert idiomatic way so I can learn it for any future changes.

Thanks, I think this is all the value I'm gonna get out of this review.  Pushing it in and moving on.
